R/cv_varsel.R

```diff
@@ -161,17 +158,21 @@ cv_varsel.vsel <- function(
     cvfits = object$cvfits,
     validate_search = object$validate_search %||% TRUE,
     ...
-) {
+    ) {
+  ## the following arguments should not change
   arg_nms_internal <- c("method", "ndraws", "nclusters", "nterms_max",
                         "search_control", "penalty", "search_terms")
   arg_nms_internal_used <- intersect(arg_nms_internal, ...names())
   n_arg_nms_internal_used <- length(arg_nms_internal_used)
-  if (n_arg_nms_internal_used > 0) {
-    stop("Argument", if (n_arg_nms_internal_used > 1) "s" else "", " ",
-         paste(paste0("`", arg_nms_internal_used, "`"), collapse = ", "), " ",
-         "cannot be specified in this case because cv_varsel.vsel() specifies ",
-         if (n_arg_nms_internal_used > 1) "them" else "it", " ", "internally.")
+  dots <- list(...)
+  for (arg in arg_nms_internal_used) {
+    if (!identical(object[[arg]], dots[[arg]])) {
```

Comment:
Does this work? I don't think all of the arguments listed in `arg_nms_internal` are accessible via `object[[arg]]`.

Reply:
Apart from that, I think I also used an `identical()` check here when starting to implement this, but I had to avoid it for some reason. I will need to think about this again.

Reply:
At least all the tests and everything else I tested works

Reply:
Here is a reprex showing that the current code does not work (as intended):
```r
data("df_gaussian", package = "projpred")
dat <- data.frame(y = df_gaussian$y, df_gaussian$x)
rfit <- rstanarm::stan_glm(y ~ X1 + X2 + X3 + X4 + X5,
                           data = dat,
                           chains = 1,
                           iter = 500,
                           seed = 1140350788,
                           refresh = 0)
devtools::load_all(".")
vs <- varsel(rfit,
             nclusters = 3,
             nclusters_pred = 5,
             nterms_max = 4,
             seed = 46782345)
cvvs <- cv_varsel(vs,
                  validate_search = FALSE,
                  nclusters = 3,
                  nclusters_pred = 5,
                  nterms_max = 4,
                  seed = 46782345)

```
throws the message
```
Argument "nclusters" ignored. Using the argument value stored in the varsel object ("").
```
Here, I guess your changes are supposed to avoid such a message. However, since `object[["nclusters"]]` does not exist (is `NULL`), the message will be thrown, with `("")` at the end.

Reply:
I think what you meant is `object[["args_search"]][[arg]]` instead of `object[[arg]]`. I just pushed commit d663df89a33a3afeabcec0c6d2f282d90fe63ec6 which fixes this. However, I think we should adapt `varsel.vsel()` correspondingly if this change is really desired. I think the reason why I picked the solution that passes arguments via the ellipsis was efficiency: Passing arguments via the ellipsis avoids the explicit evaluation of those arguments (lazy evaluation will be used instead). Creating `dots <- list(...)` will force the (early) evaluation of the arguments in `...`, which is usually undesired.

Reply:
I also realized that in the message, printing the content of the argument (in the parentheses at the end) is not possible as easily as implemented at the moment:
```r
data("df_gaussian", package = "projpred")
dat <- data.frame(y = df_gaussian$y, df_gaussian$x)
rfit <- rstanarm::stan_glm(y ~ X1 + X2 + X3 + X4 + X5,
                           data = dat,
                           chains = 1,
                           iter = 500,
                           seed = 1140350788,
                           refresh = 0)
devtools::load_all(".")
vs <- varsel(rfit,
             nclusters = 3,
             nclusters_pred = 5,
             nterms_max = 4,
             search_control = list(lambda_min_ratio = 9e-6,
                                   nlambda = 151,
                                   thresh = 9e-7),
             seed = 46782345)
cvvs <- cv_varsel(vs,
                  validate_search = FALSE,
                  nclusters = 3,
                  nclusters_pred = 5,
                  nterms_max = 4,
                  search_control = list(lambda_min_ratio = 1e-6,
                                        nlambda = 200,
                                        thresh = 1e-7),
                  seed = 46782345)

```
will throw
```
Argument "search_control" ignored. Using the argument value stored in the `vsel` object ("9e-061519e-07").
```
(in parentheses, the `list` content is merged into a single character string).

Hence, I just pushed commit d9fdbaf1ba842ae247256426505b430d49d3bba7 which removes that part.

Reply:
Concerning https://github.com/stan-dev/projpred/pull/496#discussion_r1573901190, I can't remember anymore why `identical()` should have been a problem. As I said, I think I mainly avoided syntax such as `dots <- list(...)` (and passed arguments via the ellipsis instead) for efficiency reasons.

Reply:
Resolving this for now; the unit tests should hopefully reveil any issues with this.